File: C++/surrounded-regions.cpp
class Solution {
public:
    void solve(std::vector<std::vector<char>>& board) {
        if(board.size()<3 || board[0].size()<3) return;
        int r = board.size();
        int c = board[0].size();
        std::stack<std::pair<int,int>> stk;
        for(int i=0; i<r; ++i) {
            if(board[i][0]=='O') stk.push(std::pair<int,int>(i, 0));
            if(board[i][c-1]=='O') stk.push(std::pair<int,int>(i, c-1));
        }
        
        for(int j=0; j<c; ++j) {
            if(board[0][j]=='O') stk.push(std::pair<int,int>(0, j));
            if(board[r-1][j]=='O') stk.push(std::pair<int,int>(r-1, j));
        }
        
        while(!stk.empty()) {
            int x = stk.top().first
            int y = stk.top().second;
            stk.pop();
            board[x][y] = '@';
            if(x>0 && board[x-1][y]=='O') stk.push(std::pair<int,int>(x-1, y));
            if(x+1<r && board[x+1][y]=='O') stk.push(std::pair<int,int>(x+1, y));
            if(y>0 && board[x][y-1]=='O') stk.push(std::pair<int,int>(x, y-1));
            if(y+1<r && board[x][y+1]=='O') stk.push(std::pair<int,int>(x, y+1));
            
        }
        
        for(int i=0; i<r; ++i) {
            for(int j=0; j<c; ++j) {
                if(board[i][j]=='O') board[i][j] = 'X';
                if(board[i][j]=='@') board[i][j] = 'O';
            }
        }
    }
};

File: Python/surrounded-regions.py
class Solution:
    # @param {character[][]} board
    # @return {void} Do not return anything, modify board in-place instead.
    def solve(self, board):
        if len(board)<3 or len(board[0])<3:
            return
        
        r = len(board)
        c = len(board[0])
        stk = []
        
        for i in xrange(r):
            if board[i][0]=='O': stk.append((i, 0))
            if board[i][c-1]=='O': stk.append((i, c-1))
            
        for i in xrange(c):
            if board[0][i]=='O': stk.append((0, i))
            if board[r-1][i]=='O': stk.append((r-1, i))
        
        while stk:
            i, j = stk.pop()
            board[i][j] = '@'
            if i>0 and board[i-1][j]=='O': stk.append((i-1, j))
            if i+1<r and board[i+1][j]=='O': stk.append((i+1, j))
            if j>0 and board[i][j-1]=='O': stk.append((i, j-1))
            if j+1<c and board[i][j+1]=='O': stk.append((i, j+1))
            
        for i in xrange(r):
            for j in xrange(c):
                if board[i][j]=='O': board[i][j] = 'X'
                if board[i][j]=='@': board[i][j] = 'O'
